From: Christian Kastner <ckk@kvr.at>
Date: Sun, 20 Dec 2015 19:03:04 +0100
Subject: cron: report missing newline before EOF

A missing newline before EOF caused the daemon to silently ignore the last line
of a crontab. Print a warning instead.

Bug-Debian: https://bugs.debian.org/79037
Forwarded: no
Last-Update: 2015-12-20
Index: cron/user.c
===================================================================
--- cron.orig/user.c
+++ cron/user.c
@@ -87,11 +87,20 @@ load_user(crontab_fd, pw, name)
 	 * load the crontab
 	 */
 	Set_LineNum(1)
-	while ((status = load_env(envstr, file)) >= OK) {
+	do {
+		status = load_env(envstr, file);
 		switch (status) {
 		case ERR:
-			free_user(u);
-			u = NULL;
+			/* If envstr has no content, we reached a proper EOF
+			 * and we can return to continue regular processing.
+			 *
+			 * If it does have content, we reached EOF without a
+			 * newline, so we bail out
+			 */
+			if (envstr[0] != '\0') {
+				log_it(u->name, getpid(), "ERROR", "Missing "
+				"newline before EOF");
+			}
 			goto done;
 		case FALSE:
 			e = load_entry(file, NULL, pw, envp);
@@ -110,7 +119,7 @@ load_user(crontab_fd, pw, name)
 			}
 			break;
 		}
-	}
+	} while (status >= OK);
 
  done:
 	env_free(envp);
Index: cron/entry.c
===================================================================
--- cron.orig/entry.c
+++ cron/entry.c
@@ -313,6 +313,10 @@ load_entry(file, error_func, pw, envp)
 	ch = get_string(cmd, MAX_COMMAND, file, "\n");
 
 	/* a file without a \n before the EOF is rude, so we'll complain...
+
+	   CK 2010-04-14: this code will never be reached. All calls to
+	   load_entry are proceeded by calls to load_env, which aborts on EOF, and
+	   where load_env fails, the code bails out.
 	 */
 	if (ch == EOF) {
 		ecode = e_cmd;
Index: cron/crontab.1
===================================================================
--- cron.orig/crontab.1
+++ cron/crontab.1
@@ -116,6 +116,11 @@ SVR3 syntax.
 .SH DIAGNOSTICS
 A fairly informative usage message appears if you run it with a bad command
 line.
+
+cron requires that each entry in a crontab end in a newline character. If the
+last entry in a crontab is missing the newline, cron will consider the crontab
+(at least partially) broken and refuse to install it.
+
 .SH AUTHOR
 .nf
 Paul Vixie <paul@vix.com>
Index: cron/crontab.5
===================================================================
--- cron.orig/crontab.5
+++ cron/crontab.5
@@ -321,6 +321,12 @@ the following wrapper code:
 .fi
 
 
+.SH DIAGNOSTICS
+cron requires that each entry in a crontab end in a newline character.  If the
+last entry in a crontab is missing a newline (i.e.\& terminated by EOF),
+cron will consider the crontab (at least partially) broken.
+A warning will be written to syslog.
+
 .SH AUTHOR
 .nf
 Paul Vixie <paul@vix.com>
